airbyte-featureflag/src/main/kotlin/Client.kt

```diff
@@ -0,0 +1,196 @@
+        }
+    }
+}
+
+/**
+ * Test feature-flag client. Intended only for usage in testing scenarios.
+ *
+ * All [Flag] instances will use the provided [values] map as their source of truth, including [EnvVar] flags.
+ *
+ * @param [values] is a map of [Flag.key] to enabled/disabled status.
+ */
+class TestClient(val values: Map<String, Boolean>) : FeatureFlagClient {
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> {
+                // convert to a EnvVar flag with a custom fetcher that uses the [values] of this Test class
+                // instead of fetching from the environment variables
+                EnvVar(envVar = flag.key, default = flag.default, team = flag.team) {
+                    values[flag.key]?.toString() ?: flag.default.toString()
+                }.enabled()
+            }
+
+            else -> values[flag.key] ?: flag.default
+        }
+    }
+}
+
+
+/**
+ * Data wrapper around OSS feature-flag configuration file.
+ *
+ * The file has the format of:
+ * flags:
+ *  - name: feature-one
+ *    enabled: true
+ *  - name: feature-two
+ *    enabled: false
+ */
+private data class PlatformFlags(val flags: List<PlatformFlag>)
+
+/**
+ * Data wrapper around an individual flag read from the configuration file.
+ */
+private data class PlatformFlag(val name: String, val enabled: Boolean)
+
+
+/** The yaml mapper is used for reading the feature-flag configuration file. */
+private val yamlMapper = ObjectMapper(YAMLFactory()).registerKotlinModule()
+
+/**
+ * Reads a yaml configuration file, converting it into a map of flag name to flag configuration.
+ *
+ * @param [path] to yaml config file
+ * @return map of feature-flag name to feature-flag config
+ */
+//private fun readConfig(path: Path): Map<String, OSSFlag> = yamlMapper.readValue<List<OSSFlag>>(path.toFile())
+//    .associateBy { it.name }
+private fun readConfig(path: Path): Map<String, PlatformFlag> = yamlMapper.readValue<PlatformFlags>(path.toFile())
+    .let { it.flags }
+    .associateBy { it.name }
+
+/**
+ * Monitors a [Path] for changes, calling [block] when a change is detected.
+ *
+ * @receiver Path
+ * @param [block] function called anytime a change is detected on this [Path]
+ */
+private fun Path.onChange(block: () -> Unit) {
+    val watcher = fileSystem.newWatchService()
+    // The watcher services requires a directory to be provided, hence the `parent` reference here as `this` path should be a file.
+    parent.register(watcher, StandardWatchEventKinds.ENTRY_MODIFY, StandardWatchEventKinds.ENTRY_CREATE)
+
+    thread(isDaemon = true, name = "feature-flag-watcher", priority = MIN_PRIORITY) {
+        val key = watcher.take()
+        // The context on the poll-events for ENTRY_MODIFY events should not be null, this is a sanity check more than anything else
+        key.pollEvents().mapNotNull { it.context() as? Path }
+            .filter { this.endsWith(it) }
```

Comment:
What is this filter doing, I don't understand the mix of `it` and `this`. Also in the same spirit of making the type explicit, should we avoid using `it` when we can have  a more explicit name?

Reply:
I'll add a comment to this code to make it more explicit what is going on here, as I agree you can't look at this code and know what is happening.

In Kotlin, lambdas have an [implicit lambda parameter](https://kotlinlang.org/docs/lambdas.html#it-implicit-name-of-a-single-parameter) `it` that is scoped to the lambda body.  You can provide a name to the parameter (`name ->`) but in my experience it just adds more bloat to the code with questionable benefit. Most Kotlin code uses the provided `it` parameter. 

There are situations where a named lambda parameter makes sense, such as nested lambdas, or a really long lambda block (which should be avoided).

---

The `this` in this section of code is referring to the `Path` object itself as this function is an [extension function](https://kotlinlang.org/docs/extensions.html).
```kotlin
private fun Path.onChange(block: () -> Unit) { /* ... */ }
```
This adds an `onChange` function to the `Path` object, but if this function needs to reference the `Path` object itself, it does so via `this`.  The `this.endsWith` is calling `Path.endsWith` as `this` is a `Path` type. 

Extension functions are Kotlin syntactical-sugar that essentially equate to the following java code
```
private static void onChange(final Path this, final Supplier<Void> block) { /* ... */ }
```